In [3]:
import numpy as np
import pandas as pd

# do chinh xac 2 so sau dau phay
pd.set_option("display.precision", 2)

In [4]:
df = pd.read_csv("churn.csv")
df.head(3)

State  Account Length  Area Code     Phone Int'l Plan VMail Plan  \
0    KS             128        415  382-4657         no        yes   
1    OH             107        415  371-7191         no        yes   
2    NJ             137        415  358-1921         no         no   

   VMail Message  Day Mins  Day Calls  Day Charge  ...  Eve Calls  Eve Charge  \
0             25     265.1        110       45.07  ...         99       16.78   
1             26     161.6        123       27.47  ...        103       16.62   
2              0     243.4        114       41.38  ...        110       10.30   

   Night Mins  Night Calls  Night Charge  Intl Mins  Intl Calls  Intl Charge  \
0       244.7           91         11.01       10.0           3         2.70   
1       254.4          103         11.45       13.7           3         3.70   
2       162.6          104          7.32       12.2           5         3.29   

   CustServ Calls  Churn?  
0               1  False.  
1               1  False.  
2               0  False.  

[3 rows x 21 columns]

About printing DataFrames in Jupyter notebooks

Recall that each row corresponds to one client, an instance, and columns
are features of this instance

Let's have a look at data dimensionality, feature names and feature types

In [5]:
print(df.shape)

(3333, 21)


From the ouput we can see that the table contains 3333 rows and 21 columns.

Now let's try printing out column names using `columns`

In [6]:
df.columns

Index(['State', 'Account Length', 'Area Code', 'Phone', 'Int'l Plan',
       'VMail Plan', 'VMail Message', 'Day Mins', 'Day Calls', 'Day Charge',
       'Eve Mins', 'Eve Calls', 'Eve Charge', 'Night Mins', 'Night Calls',
       'Night Charge', 'Intl Mins', 'Intl Calls', 'Intl Charge',
       'CustServ Calls', 'Churn?'],
      dtype='object')

We can use info() method to output some general information about the dataframe:

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   State           3333 non-null   object 
 1   Account Length  3333 non-null   int64  
 2   Area Code       3333 non-null   int64  
 3   Phone           3333 non-null   object 
 4   Int'l Plan      3333 non-null   object 
 5   VMail Plan      3333 non-null   object 
 6   VMail Message   3333 non-null   int64  
 7   Day Mins        3333 non-null   float64
 8   Day Calls       3333 non-null   int64  
 9   Day Charge      3333 non-null   float64
 10  Eve Mins        3333 non-null   float64
 11  Eve Calls       3333 non-null   int64  
 12  Eve Charge      3333 non-null   float64
 13  Night Mins      3333 non-null   float64
 14  Night Calls     3333 non-null   int64  
 15  Night Charge    3333 non-null   float64
 16  Intl Mins       3333 non-null   float64
 17  Intl Calls      3333 non-null   i

`bool, int64, float64` and `object` are data types of our features. We see that one feature is logical (`bool`), 3 features are of type `object`, and 16 features are numericc. With this same method, we can easly see if there are any missing values. Here, there are none because each column contains 3333 observations, the same number of rows we saw before with shape

We can change the column type with the `astype` method. Let's apply this method to the `Churn` feature to convert it into  `bool` then `int64`



In [8]:
d = {"False.": False, "True.": True}
df['Churn?'] = df['Churn?'].map(d)

In [11]:
# after convert from str to bool, let's continue convert `bool` to `int64`

df['Churn?'] = df['Churn?'].astype('int64')

The `describe()` method shows basic statistical characteristics of each numerical feature ( `int64` and `float64` types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles

In [12]:
df.describe()

Account Length  Area Code  VMail Message  Day Mins  Day Calls  \
count         3333.00    3333.00        3333.00   3333.00    3333.00   
mean           101.06     437.18           8.10    179.78     100.44   
std             39.82      42.37          13.69     54.47      20.07   
min              1.00     408.00           0.00      0.00       0.00   
25%             74.00     408.00           0.00    143.70      87.00   
50%            101.00     415.00           0.00    179.40     101.00   
75%            127.00     510.00          20.00    216.40     114.00   
max            243.00     510.00          51.00    350.80     165.00   

       Day Charge  Eve Mins  Eve Calls  Eve Charge  Night Mins  Night Calls  \
count     3333.00   3333.00    3333.00     3333.00     3333.00      3333.00   
mean        30.56    200.98     100.11       17.08      200.87       100.11   
std          9.26     50.71      19.92        4.31       50.57        19.57   
min          0.00      0.00       0.00        0.00       23.20        33.00   
25%         24.43    166.60      87.00       14.16      167.00        87.00   
50%         30.50    201.40     100.00       17.12      201.20       100.00   
75%         36.79    235.30     114.00       20.00      235.30       113.00   
max         59.64    363.70     170.00       30.91      395.00       175.00   

       Night Charge  Intl Mins  Intl Calls  Intl Charge  CustServ Calls  \
count       3333.00    3333.00     3333.00      3333.00         3333.00   
mean           9.04      10.24        4.48         2.76            1.56   
std            2.28       2.79        2.46         0.75            1.32   
min            1.04       0.00        0.00         0.00            0.00   
25%            7.52       8.50        3.00         2.30            1.00   
50%            9.05      10.30        4.00         2.78            1.00   
75%           10.59      12.10        6.00         3.27            2.00   
max           17.77      20.00       20.00         5.40            9.00   

        Churn?  
count  3333.00  
mean      0.14  
std       0.35  
min       0.00  
25%       0.00  
50%       0.00  
75%       0.00  
max       1.00

In order to see statistics on non-numerical features, one has to explicitly(rõ ràng) indicate data types of interest in the `inculde` parameter.

In [13]:
df.describe(include=['bool', 'object'])

State     Phone Int'l Plan VMail Plan
count   3333      3333       3333       3333
unique    51      3333          2          2
top       WV  353-6218         no         no
freq     106         1       3010       2411

For categorical (type `object`) and boolen (type `bool`) features we can use the `value_counts` method. Let's have a look at the distribution of `Churn`

In [14]:
df["Churn?"].value_counts()

0    2850
1     483
Name: Churn?, dtype: int64

as we can see 2850 users out of 3333 are loyal; their `Churn` value is `0`. To caculate factions, pass `normalize=True` to the `value_counts` function.

In [15]:
df['Churn?'].value_counts(normalize=True)

0    0.86
1    0.14
Name: Churn?, dtype: float64